src/toil/jobStores/abstractJobStore.py

```diff
@@ -520,47 +520,67 @@ def getJobDescriptions():
             else:
                 return self.jobs()
 
-        # Iterate from the root JobDescription and collate all jobs that are reachable from it
-        # All other jobs returned by self.jobs() are orphaned and can be removed
-        reachableFromRoot = set()
-
-        def getConnectedJobs(jobDescription):
-            if jobDescription.jobStoreID in reachableFromRoot:
-                return
-            reachableFromRoot.add(jobDescription.jobStoreID)
-            # Traverse jobs in stack
-            for jobs in jobDescription.stack:
-                for successorJobStoreID in jobs:
-                    if (successorJobStoreID not in reachableFromRoot
-                        and haveJob(successorJobStoreID)):
-                        getConnectedJobs(getJobDescription(successorJobStoreID))
-            # Traverse service jobs
-            for serviceJobStoreID in jobDescription.services:
-                if haveJob(serviceJobStoreID):
-                    assert serviceJobStoreID not in reachableFromRoot
-                    reachableFromRoot.add(serviceJobStoreID)
-
-        logger.debug("Checking job graph connectivity...")
-        getConnectedJobs(self.loadRootJob())
-        logger.debug("%d jobs reachable from root." % len(reachableFromRoot))
+        def get_jobs_reachable_from_root() -> Set[Union[TemporaryID, str]]:
+            """
+            Traverse the job graph from the root job and return a flattened set of all active jobstore IDs.
+
+            Note: Jobs returned by self.jobs(), but not this function, are orphaned, and can be removed as dead jobs.
+            """
+            def successors(job_descriptions: List[JobDescription]) -> Iterable[JobDescription]:
+                """
+                For a list of job description objects, this will iterate each job description object's stack
+                of connected jobs and yield any that have both an active job and have not yet been added to
+                reachable_from_root.
+                """
+                for job_description in job_descriptions:
+                    for jobs in job_description.stack:
+                        for successor_jobstore_id in jobs:
+                            if successor_jobstore_id not in reachable_from_root and haveJob(successor_jobstore_id):
```

Comment:
Since reachable_from_root is a set. What is the purpose of the membership check?

Set lookup is similar complexity to insertion, so you can save half the work by skipping the membership check (which will probably be meaningless to Toil's overall performance).

Reply:
Ah, it's that it adds the job id plus all attached service ids if it's not in the set.  I can see about replacing the check for a length check after adding to see if that's more efficient.  This graph can get large (at least enough to break recursion, but possibly in the millions, though much more often less than 100), so gains in efficiency here might be noticeable for particularly large workflows.

Thanks for taking a look!  :slightly_smiling_face: